MuonAnalysis/MuonAssociators/src/PropagateToMuonCore.cc

```diff
@@ -1,72 +1,46 @@
-                                                                 : false),
-      magfieldToken_(iC.esConsumes<>()),
-      propagatorToken_(iC.esConsumes<>(edm::ESInputTag("", "SteppingHelixPropagatorAlong"))),
-      propagatorAnyToken_(iC.esConsumes<>(edm::ESInputTag("", "SteppingHelixPropagatorAny"))),
-      propagatorOppositeToken_(iC.esConsumes<>(edm::ESInputTag("", "SteppingHelixPropagatorOpposite"))),
-      muonGeometryToken_(iC.esConsumes<>()) {
-  std::string whichTrack = iConfig.getParameter<std::string>("useTrack");
-  if (whichTrack == "none") {
-    whichTrack_ = None;
-  } else if (whichTrack == "tracker") {
-    whichTrack_ = TrackerTk;
-  } else if (whichTrack == "muon") {
-    whichTrack_ = MuonTk;
-  } else if (whichTrack == "global") {
-    whichTrack_ = GlobalTk;
-  } else
-    throw cms::Exception("Configuration") << "Parameter 'useTrack' must be 'none', 'tracker', 'muon', 'global'\n";
-  if (whichTrack_ != None) {
-    std::string whichState = iConfig.getParameter<std::string>("useState");
-    if (whichState == "atVertex") {
-      whichState_ = AtVertex;
-    } else if (whichState == "innermost") {
-      whichState_ = Innermost;
-    } else if (whichState == "outermost") {
-      whichState_ = Outermost;
-    } else
-      throw cms::Exception("Configuration") << "Parameter 'useState' must be 'atVertex', 'innermost', 'outermost'\n";
-  }
-  if (cosmicPropagation_ && (whichTrack_ == None || whichState_ == AtVertex)) {
-    throw cms::Exception("Configuration") << "When using 'cosmicPropagationHypothesis' useTrack must not be 'none', "
-                                             "and the state must not be 'atVertex'\n";
-  }
-}
-
-PropagateToMuon::~PropagateToMuon() {}
-
-void PropagateToMuon::init(const edm::EventSetup &iSetup) {
-  magfield_ = iSetup.getHandle(magfieldToken_);
-  propagator_ = iSetup.getHandle(propagatorToken_);
-  propagatorOpposite_ = iSetup.getHandle(propagatorOppositeToken_);
-  propagatorAny_ = iSetup.getHandle(propagatorAnyToken_);
-  muonGeometry_ = iSetup.getHandle(muonGeometryToken_);
-
+PropagateToMuonCore::PropagateToMuonCore(){};
+
+PropagateToMuonCore::PropagateToMuonCore(edm::ESHandle<MagneticField> magfield,
+                                         edm::ESHandle<Propagator> propagator,
+                                         edm::ESHandle<Propagator> propagatorAny,
+                                         edm::ESHandle<Propagator> propagatorOpposite,
+                                         edm::ESHandle<MuonDetLayerGeometry> muonGeometry,
+                                         bool useSimpleGeometry,
+                                         bool useMB2,
+                                         bool fallbackToME1,
+                                         WhichTrack whichTrack,
+                                         WhichState whichState,
+                                         bool cosmicPropagation,
+                                         bool useMB2InOverlap)
```

Comment:
Looks like all args could be made `const`.

Reply:
This is not possible at the moment as we still need the empty constructor where they are not initialized.